compiler/src/iree/compiler/Codegen/Common/ExtractAddressComputation.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/ExtractAddressComputation.h"
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "extract-address-computation"
+
+using namespace mlir;
+
+namespace mlir {
+namespace iree_compiler {
+
+//===----------------------------------------------------------------------===//
+// Helper functions for the `load base[off0...]`
+//  => `load (subview base[off0...])[0...]` pattern.
+//===----------------------------------------------------------------------===//
+
+// Matches getSrcMemRef specs for LoadOp.
+// \see LoadLikeOpRewriter.
+static Value getLoadOpSrcMemRef(memref::LoadOp loadOp) {
+  return loadOp.getMemRef();
+}
+
+// Matches rebuildOpFromAddressAndIndices specs for LoadOp.
+// \see LoadLikeOpRewriter.
+static memref::LoadOp rebuildLoadOp(RewriterBase &rewriter,
+                                    memref::LoadOp loadOp, Value srcMemRef,
+                                    ArrayRef<Value> indices) {
+  Location loc = loadOp.getLoc();
+  return rewriter.create<memref::LoadOp>(loc, srcMemRef, indices);
+}
+
+SmallVector<OpFoldResult> getLoadOpViewSizeForEachDim(RewriterBase &rewriter,
+                                                      memref::LoadOp loadOp) {
+  MemRefType ldTy = loadOp.getMemRefType();
+  unsigned loadRank = ldTy.getRank();
+  return SmallVector<OpFoldResult>(loadRank, rewriter.getIndexAttr(1));
+}
+
+void populateExtractAddressComputationPatterns(RewritePatternSet &patterns) {
+  patterns.add<LoadLikeOpRewriter<
+      memref::LoadOp,
+      /*getSrcMemRef=*/getLoadOpSrcMemRef,
+      /*rebuildOpFromAddressAndIndices=*/rebuildLoadOp,
+      /*getViewSizeForEachDim=*/getLoadOpViewSizeForEachDim>>(
+      patterns.getContext());
+}
+
+//===----------------------------------------------------------------------===//
+// Pass registration
+//===----------------------------------------------------------------------===//
+namespace {
+
+struct ExtractAddressComputationPass
+    : public ExtractAddressComputationBase<ExtractAddressComputationPass> {
+  void runOnOperation() override;
+};
+}  // namespace
+
+void ExtractAddressComputationPass::runOnOperation() {
+  RewritePatternSet patterns(&getContext());
+  populateExtractAddressComputationPatterns(patterns);
+  (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
```

Comment:
This would be good to catch and signal failure. This typically returns failure when pattern application fails to converge, which should be an error and better to catch that.

Reply:
Good point.
I was only thinking in terms of optimization: this optimization is optional so if it fails we don't care, but actually we may not be converging like you said (albeit it cannot happen right now :)).

I'll fix that.